cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -818,9 +830,20 @@ private synchronized boolean stopExecution(boolean forceExecutionStop) {
    */
   public synchronized void shutdown() {
     LOG.info("Shutting down executor.");
+
+    _flipOngoingExecutionMutex.acquireUninterruptibly();
```

Comment:
Thanks Efe for the snippet! Can you tell me more on why we need 3 `try` here? When I rewrite my code I was thinking about writing it like this:
```
    try {
      _flipOngoingExecutionMutex.acquire();
      if (_hasOngoingExecution) {
        LOG.warn("Shutdown executor may take long because execution is still in progress.");
        stopExecution(false);
      }
      try {
        // Wait until ongoing execution stopped completely. Block new execution from proceeding.
        _noOngoingExecutionSemaphore.acquire();
      } catch (final InterruptedException ie) {
        LOG.warn("Interrupted while waiting for ongoing execution to stop.");
      }
    } catch (final InterruptedException ie) {
      // handle acquire failure for _flipOngoingExecutionMutex here
      LOG.warn("Interrupted while checking if there is ongoing execution to stop.");
    }
    finally {
      _flipOngoingExecutionMutex.release();
    }
```

Reply:
@zornhsu An `InterruptedException` would mean that no permit has been acquired and no `release()` should be done. However, with your proposed approach, if `_flipOngoingExecutionMutex.acquire();` throws an `InterruptedException`, you release `_flipOngoingExecutionMutex` incorrectly. That is why we need 3 `try`.
Let me know if you'd have further questions!

Reply:
Thanks, @efeg. I think I know why my previous approach was wrong. 
Can you tell me more on why we need the 2nd `try`? I am not sure what would go wrong if I omitting it like this:
```java
    try {
      _flipOngoingExecutionMutex.acquire();
      if (_hasOngoingExecution) {
        LOG.warn("Shutdown executor may take long because execution is still in progress.");
        stopExecution(false);
      }
      try {
        // Wait until ongoing execution stopped completely. Block new execution from proceeding.
        _noOngoingExecutionSemaphore.acquire();
      } catch (final InterruptedException ie) {
        LOG.warn("Interrupted while waiting for ongoing execution to stop.");
      }
      _flipOngoingExecutionMutex.release();
    } catch (final InterruptedException ie) {
      // handle acquire failure for _flipOngoingExecutionMutex here
      LOG.warn("Interrupted while checking if there is ongoing execution to stop.");
    }
```
Is it because we might add more code in between in the future which might throw other exceptions, preventing us reaching `release()`? So it's a better practice to use this extra 2nd `try`?

Reply:
Synced with Efe offline. The reason to add the 2nd `try` is the same as what I guessed above.